[GO] Rework go language support (1)

oakad opened the pull request:
As a follow up to issue #5072 I have reworked the go support component in the flatc. This is very much a work in progress, however I believe it may be useful to get some feedback at this stage.

On a pro side:
1. The new implementation uses the CodeWriter and is generally refactored to be a class rather than collection of functions.
2. "Create" message function for tables
2. Make builder into a proper object with methods
3. Fix the long standing import bug
4. Preserve the enum types whenever possible for type safety
5. Native api generation (--gen-object-api) - in progress 

On a cons side:
1. Not formatted according to the google code guidelines
2. Changes the Go API and semantics of a command line flag "--go-namespace"
3. Has no "single file" mode - in Go, namespaces are tied tightly to folders, so most, at least marginally complex, schemas would require multiple files anyway.

Signed-off-by: Alex Dubov <oakad@yahoo.com>

oakad commented:
I signed it

oakad commented on tests/go_test.go:
```diff
@@ -152,7 +152,7 @@ func CheckReadBuffer(buf []byte, offset flatbuffers.UOffsetT, fail func(string,
 			fail(FailString("mana", 150, got))
```
I am breaking the API, that's the whole idea. The present API is bordering on unusable outright (and my schema is not even that complex). 

The same applies to the generated code, which is also quite different.

My problem is, I want to use (for now, at least) flatbuffers to interface with preexisting c++ stuff. That's why I can't take "baby steps" on this.

I do appreciate that the interim result will probably be non-mergeable from your perspective. However, if there's an interest in FBs on Go (there was not so much until now), this PR may be a useful place to discuss the more appropriate changes to the main FB library.

aardappel replied:
Hmm.. While I am sure we can improve a lot about the API, I am not we'd want to break all of it. And maintaining 2 side by side generators (and test/sample code, etc etc) sounds pretty bad too. I defer to @rw on this.

oakad replied:
Having multiple implementations of whatever language is not the most desirable outcome, yes, but it is definitely not unheard of. After all, IDL libraries are just a glue for the much larger applications, and those application may exhibit a rather different trade off profiles, especially when managed memory languages are concerned, where direct casting of byte ranges to application values is not available.

rw commented on tests/go_test.go:
```diff
@@ -152,7 +152,7 @@ func CheckReadBuffer(buf []byte, offset flatbuffers.UOffsetT, fail func(string,
 			fail(FailString("mana", 150, got))
 		}
 
-		if got := monster.Name(); !bytes.Equal([]byte("MyMonster"), got) {
```
@oakad Not quite true: `string`s require heap allocations and byte slices don't. (Unless we use the `unsafe` functionality.)

oakad replied:
@rw Of course they require it. However, if we look at the application level, byte slices are usually a not a substitute for strings anywhere apart from C/C++. So yes, the IDL library can pretend to save a bit of effort not doing that allocation, but the library user will have to construct the string first thing upon returning from the call, with that allocation and all.

The performance/allocation savvy users can always use `[byte]` in critical places, but considering that `string` is probably the most used (and by a big margin) data type in Go API's they are not going to get too far with it.

rw commented:
@oakad Wow, this is a lot of effort! Thanks for thinking through this and getting into the code.

So that we can discuss this with more feedback, would you be able to come up with a design document for the changes you want to make? Ideally, you'd have sections that state `before` and `after` for the changes you are proposing. As-is, it's not clear how to evaluate the design decisions apart from the code changes.

oakad commented:
I will write a description when I get a minute or two. Basically, instead of using free constructor functions I have introduced a builder object (in line with most other languages) and implemented the native object marshaling code generation.

oakad commented:
https://github.com/oakad/flatbuffers/blob/master/docs/source/GoApi_new.md

aardappel commented:
@oakad 
Another change you are making is "builder objects", which causes an object to be allocated for each object constructed. The whole point of FlatBuffers is to get away from allocating objects and keep things as efficient as possible.

You are also introducing an object API, so why not reserve nice convenient API features for that (since it is meant for uses where convenience is more important than maximum speed) and leave the base API biased towards speed over convenience? In fact, if you go that way, you can keep @rw's base API, and simply make the object API use it.. that's how it works in C++.

oakad commented:
@aardappel 

> which causes an object to be allocated for each object constructed

This is incorrect. Most of the time those objects are simply pushed to stack (it's not Java here, Go has a well defined concept of "automatic" variables as well as pointers to stack held vars).

I would also like to point out that "reducing the number of allocations" while desirable, should not become a fetish in memory managed languages (in early Go versions memory allocator was very so-so, but it has improved dramatically since then). For example, an allocation on average may take less time then casting a byte range to a float (and it may even take less time than casting a byte range to an int) - because most of the time an allocation has an amortized constant time complexity comparable to a stack push. Consider protobufs for example - that library allocates a massive amount of short lived objects and still manages to pull of a rather decent performance in Go.

As to the keeping the old API - it confuses the IDE autocomplete enough to make it very annoying to work (at least with my schema). But this can be addressed by a flag, I suppose.

Because the new compiler implementation is more or less a complete rewrite I will have to explicitly add some code for it.

aardappel commented:
@oakad 
> This is incorrect. Most of the time those objects are simply pushed to stack
In the case of Java I am aware of e.g. Hotspot doing "escape analysis" to achieve this, but I am not sure that every implementation does this, for example Android, where efficiency is the most important.
In the case of Go, can you point me to a list or rules that guarantees that in this case this optimization will happen?

> I would also like to point out that "reducing the number of allocations" while desirable, should not become a fetish in memory managed languages
I am not interested in starting another "the cost of GC" discussion here again. Fact is that the whole point of FlatBuffers is to replace objects by a nice compact byte array. Whether that is an optimization worth having probably depends on the language and use case, but should not be thrown away by default. I have nothing against an object API, but I think it should be the choice of the programmer to opt-in to it.

oakad commented:
It's not an optimization, it's language model. In Go, neither creating a struct nor passing structs by value causes any allocations. Consider the following program (https://godbolt.org/z/PW4WwJ):

```
package main

import (
	"fmt"
)

type Hello struct {
    Value string
}

func print(a Hello) {
	fmt.Println(a)
}

func main() {
	a := Hello {
		Value: "Yeah, baby!",
	}
	print(a)
}
```
No allocations will happen here apart from `runtime.convT2E` before the call to `fmt.Println`.

oakad commented:
I have brought back the original builder API, even though I think it is way to cumbersome for any practical use.

The flatbuffer strings are still treated as Go strings unconditionally, because pretending that "we can avoid that allocation by using a byte slice instead" is merely a fig leave, hiding no naked truth. :-)

aardappel commented:
@binary132 thanks, that's what I thought.

@oakad people use FlatBuffers because (part of) their tech stack requires certain performance characteristics. I am sure they can handle a slightly less pretty looking API in exchange for that, as many have in languages other than Go.

As for strings, I suggest to simply allow both.. you can default to strings if you wish, but a parallel byte slice accessor would be good.

I suggest benchmarking against @rw's existing API to ensure we're not losing any past performance. There is already a benchmark I believe.

oakad commented:
@aardappel  @binary123's remark has nothing to do with the issue at hand, because escape analysis is not applicable to structs passed by value. In Go, not every value is a reference - it's not Java by any measure.

Byte arrays are already allowed: people can just use `[ubyte]` in schema if they feel like it. Adding a switch to a compiler result in a horrible mess of code for no good reason (I've actually tried it already).

As to the benchmark (the only applicable one is `BenchmarkBuildGold-4`) it ends up exactly the same (and I imagine the generated machine code is also quite similar).

Original (after tweaking for parent name space issue):

BenchmarkBuildGold-4                     5000000               957 ns/op         233.91 MB/s            0 B/op          0 allocs/op

Mine:
BenchmarkBuildGold-4                     5000000               947 ns/op         236.48 MB/s            0 B/op          0 allocs/op

````
// BenchmarkBuildGold uses generated code to build the example Monster.
func BenchmarkBuildGold(b *testing.B) {
        buf, offset := CheckGeneratedBuild(b.Fatalf)
        bytes_length := int64(len(buf[offset:]))

        reuse_str := "MyMonster"
        reuse_test1 := "test1"
        reuse_test2 := "test2"
        reuse_fred := "Fred"

        b.SetBytes(bytes_length)
        bldr := example.BuildMonster(flatbuffers.NewBuilder(0))
        b.ResetTimer()
        b.ReportAllocs()
        for i := 0; i < b.N; i++ {
                bldr.Reset()

                str := bldr.CreateString(reuse_str)
                test1 := bldr.CreateString(reuse_test1)
                test2 := bldr.CreateString(reuse_test2)
                fred := bldr.CreateString(reuse_fred)

                bldr.StartInventoryVector(5)
                bldr.PrependByte(4)
                bldr.PrependByte(3)
                bldr.PrependByte(2)
                bldr.PrependByte(1)
                bldr.PrependByte(0)
                inv := bldr.EndVector(5)

                bldr.Start()
                bldr.AddName(fred)
                mon2 := bldr.End()

                bldr.StartTest4Vector(2)
                example.CreateTest(bldr.Builder, 10, 20)
                example.CreateTest(bldr.Builder, 30, 40)
                test4 := bldr.EndVector(2)

                bldr.StartTestarrayofstringVector(2)
                bldr.PrependUOffsetT(test2)
                bldr.PrependUOffsetT(test1)
                testArrayOfString := bldr.EndVector(2)

                bldr.Start()

                pos := example.CreateVec3(bldr.Builder, 1.0, 2.0, 3.0, 3.0, 2, 5, 6)
                bldr.AddPos(pos)

                bldr.AddHp(80)
                bldr.AddName(str)
                bldr.AddInventory(inv)
                bldr.AddTestType(1)
                bldr.AddTest(mon2)
                bldr.AddTest4(test4)
                bldr.AddTestarrayofstring(testArrayOfString)
                bldr.Finish(bldr.End())
        }
}
````

rw commented:
1) We merged #5097 ([Go] Namespaced imports fix) which is relevant here.

2) I agree with @aardappel that allocations are to be avoided. In particular, even if the Go spec doesn't provide strict rules, we know how the most popular toolchain behaves w.r.t. escape analysis. There's no reason to ignore that information.

3) It is still true that safely reading a `string` from a FlatBuffer datum requires an allocation in Go, and it is still true that safely reading a `[]byte` does not. (Note that we have had code that uses `unsafe` to read `string` data with no allocation.)

oakad commented:
@rw

1. Right

2. I have not introduced any new allocations, apart from the object api

3. You can not construct a valid go application without paying for string or slice field value allocation. Because, either you reuse the main buffer for the message -  this forces allocation for a field value copy (no matter if string or slice) or you don't reuse the main buffer, so you need to allocate it anew for each message.

At any rate, I shall not pursue any further discussions on this topic. Feel free to drop this PR.

aardappel commented:
@oakad if you're not willing to make compromises on your PR we are requesting, then yes we can't merge it.

rw commented:
@oakad Could we interest you in working on some important, specific contributions? Go still needs API improvements, we just want to write them in a more piecemeal (and consensus-driven) way.

1) A "Builder API" like what is in Rust and C++:
https://github.com/google/flatbuffers/blob/master/tests/monster_test_generated.rs#L940
https://github.com/google/flatbuffers/blob/master/tests/monster_test_generated.rs#L1343

2) An "Object API" like what is in C++:
https://github.com/google/flatbuffers/blob/master/tests/monster_test_generated.h#L1015

I'm happy to guide / review anything you want related to those.

aardappel commented:
Closing.. if anyone wants to continue this work, please open a new PR that attempts to address the issues raised above.
